### PR TITLE
feat: add disable_proposal_forwarding config params

### LINKED ISSUE #550: Feat: add raft config parameters `disable_proposal_forwarding`
**Is your feature request related to a problem? Please describe.**
[etcd raft](https://github.com/etcd-io/raft/blob/d8715c5aecc21ab7c539b30f70c3bd7696e5bce0/raft.go#L256C2-L256C27) provides a config param `DisableProposalForwarding` which is useful in case the computation to create proposal must be performed on leader node before added to `raft log` and replicated.

**Describe the solution you'd like**
- Add a configuration option `disable_proposal_forwarding` with default value is `false`.
```rust
pub struct Config {
    ...
    pub disable_proposal_forwarding: bool,
    ...
}
```
- In `step_follower`, forward a proposal if there is a valid leader and `disable_proposal_forward = false`, otherwise return `Error::ProposalDropped`.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by datbeohbbh**
resolve #550

### CODE REVIEW

File: src/raft.rs [hunk 1]
```diff
@@ -255,6 +255,8 @@ pub struct RaftCore<T: Storage> {
 
     /// Max size per committed entries in a `Read`.
     pub(crate) max_committed_size_per_ready: u64,
+
+    disable_proposal_forwarding: bool,
```
Comment (BusyJay) [inline src/raft.rs:259]: How about put it near other bool definition?

File: harness/tests/integration_cases/test_raft.rs [hunk 2]
```diff
@@ -5851,3 +5851,123 @@ fn test_switching_check_quorum() {
     }
     assert_eq!(sm.state, StateRole::Leader);
 }
+
+#[test]
+fn test_disable_proposal_forwarding() {
```
Comment (BusyJay) [inline harness/tests/integration_cases/test_raft.rs:5856]: Better follow the same test case as etcd. https://github.com/etcd-io/raft/blob/d6c02139d1de7eb4bea79a266ea94bab6a4e0864/node_test.go#L166-L192

Comment (datbeohbbh) [reply harness/tests/integration_cases/test_raft.rs:5856]: ok, updated

Comment (BusyJay) [reply harness/tests/integration_cases/test_raft.rs:5856]: Can you also move the test case to test_raw_node.rs?

Comment (BusyJay) [reply harness/tests/integration_cases/test_raft.rs:5856]: We align test cases to etcd. Test cases from raft_test.go are implemented in test_raft.rs, from node_test.go are implemented in test_raw_node.rs.

Comment (datbeohbbh) [reply harness/tests/integration_cases/test_raft.rs]: ok, I moved the test case to raw node test.

**Comment by datbeohbbh** [comment]
@BusyJay can you please check CI clippy? It's seem that clippy throw error on other code parts which are not relevant to this PR.

**Comment by BusyJay** [comment]
If you are interested, you can fix them in this PR (or a new PR).

**Comment by datbeohbbh** [comment]
> If you are interested, you can fix them in this PR (or a new PR).

I will fix in my next PR. This PR should be about this feature :D

**Comment by BusyJay** [comment]
> > If you are interested, you can fix them in this PR (or a new PR).
> 
> I will fix in my next PR. This PR should be about this feature :D

OK, let's hold this one for now till CI is working again.

**Comment by datbeohbbh** [comment]
> > > If you are interested, you can fix them in this PR (or a new PR).
> > 
> > 
> > I will fix in my next PR. This PR should be about this feature :D
> 
> OK, let's hold this one for now till CI is working again.
@BusyJay CI worked, I have rebased this PR onto `master`.

**Review by BusyJay — APPROVED** [review]

**Comment by BusyJay** [comment]
Well done! Thank you!

**Pull request merged**